src/main/java/com/hms_networks/sc/cumulocity/api/CConnectorMqttMgr.java

```diff
@@ -0,0 +1,376 @@
+  private int mqttLoopAttemptCount = 0;
+
+  /** MQTT loop limit, before triggering a reprovision event */
+  private int mqttLoopAttemptLimit = 15;
+
+  /** Boolean flag to track subscribing to the required MQTT topics. */
+  private boolean subscribedToTopics = false;
+
+  /** String containing the ID of the MQTT client. */
+  private final String mqttId;
+
+  /**
+   * Constructor for the MQTT manager with the given MQTT ID, host, and boolean indicating if UTF-8
+   * support is enabled.
+   *
+   * @param mqttId The ID of the MQTT client.
+   * @param mqttHost The host of the MQTT broker.
+   * @param enableUtf8 Boolean indicating if UTF-8 support is enabled.
+   * @throws Exception if unable to create the MQTT client.
+   */
+  public CConnectorMqttMgr(String mqttId, String mqttHost, boolean enableUtf8) throws Exception {
+    super(mqttId, mqttHost, enableUtf8);
+    this.mqttId = mqttId;
+
+    // Configure desired MQTT loop sleep interval
+    setMqttThreadSleepIntervalMs(MQTT_LOOP_WAIT_MILLIS);
+
+    // Add MQTT listener to default event handler
+    DefaultEventHandler.addMqttListener(this);
+
+    // Add alarm listener to default event handler
+    DefaultEventHandler.setDefaultTagAlarmListener(new CConnectorAlarmMgr());
+  }
+
+  /**
+   * Method for managing errors (exceptions) that occur in the MQTT manager.
+   *
+   * @param throwable exception to be handled
+   */
+  public void onError(Throwable throwable) {
+    Logger.LOG_SERIOUS("The MQTT client has encountered an error (See exception details)!");
+    Logger.LOG_EXCEPTION((Exception) throwable);
+  }
+
+  /**
+   * Method for managing received MQTT messages.
+   *
+   * @param mqttMessage received MQTT message
+   */
+  public void onMessage(MqttMessage mqttMessage) {
+    // Pass message to API message reader/parser
+    String mqttMessagePayload = new String(mqttMessage.getPayload());
+    CConnectorApiMessageReader.parseMessage(
+        this, mqttMessage.getTopic(), mqttMessagePayload, mqttId);
+  }
+
+  /** Method for processing or performing tasks on the looping MQTT thread. */
+  public void runOnMqttLoop() {
+    try {
+      if (getStatus() != MqttStatusCode.CONNECTED) {
+        // Increment loop attempt count
+        mqttLoopAttemptCount++;
+        if (mqttLoopAttemptCount > mqttLoopAttemptLimit) {
+          Logger.LOG_CRITICAL(
+              "MQTT disconnected for "
+                  + mqttLoopAttemptCount
+                  + " cycles, triggering a reprovision.");
+          CConnectorMain.rerunProvisioning();
+        }
+      }
+    } catch (EWException ew) {
+      Logger.LOG_SERIOUS("Exception getting MQTT status!");
+      Logger.LOG_EXCEPTION(ew);
+    }
+  }
+
+  /**
+   * Method for managing MQTT status changes.
+   *
+   * @param status MQTT status code
+   */
+  public void onStatus(int status) {
+    Logger.LOG_CRITICAL("MQTT client status changed to " + status);
```

Comment:
it's strange to me that the rest of the application is not aware of the MQTT state. Regardless of state, it does the same thing.